core/types/meta.go

```diff
@@ -0,0 +1,100 @@
+/**
+ * Optimism 2020 Copyright
+ */
+
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+type TransactionMeta struct {
+	L1RollupTxId      *hexutil.Uint64    `json:"l1RollupTxId"`
+	L1MessageSender   *common.Address    `json:"l1MessageSender"`
+	SignatureHashType *SignatureHashType `json:"signatureHashType"`
+}
+
+func NewTransactionMeta(L1RollupTxId *hexutil.Uint64, L1MessageSender *common.Address, sighashType *SignatureHashType) *TransactionMeta {
+	return &TransactionMeta{L1RollupTxId: L1RollupTxId, L1MessageSender: L1MessageSender, SignatureHashType: sighashType}
+}
+
+// TxMetaDecode deserializes bytes as a TransactionMeta struct.
+// The schema is:
+// varbytes(SignatureHashType) || varbytes(L1RollupTxId) || varbytes(L1MessageSender)
+func TxMetaDecode(input []byte) (*TransactionMeta, error) {
+	var err error
+	var meta TransactionMeta
+	b := bytes.NewReader(input)
+
+	sb, err := common.ReadVarBytes(b, 0, 1024, "SignatureHashType")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	var sighashType SignatureHashType
+	binary.Read(bytes.NewReader(sb), binary.LittleEndian, &sighashType)
+	meta.SignatureHashType = &sighashType
+
+	lb, err := common.ReadVarBytes(b, 0, 1024, "L1RollupTxId")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	if !bytes.Equal(lb, []byte{0x00}) {
+		var l1RollupTxId hexutil.Uint64
+		binary.Read(bytes.NewReader(lb), binary.LittleEndian, &l1RollupTxId)
+		meta.L1RollupTxId = &l1RollupTxId
+	}
+
+	mb, err := common.ReadVarBytes(b, 0, 1024, "L1MessageSender")
+	if err != nil {
+		return &TransactionMeta{}, err
+	}
+
+	if !bytes.Equal(mb, []byte{0x00}) {
+		var l1MessageSender common.Address
+		binary.Read(bytes.NewReader(mb), binary.LittleEndian, &l1MessageSender)
+		meta.L1MessageSender = &l1MessageSender
+	}
+
+	return &meta, nil
+}
+
+// TxMetaEncode serializes the TransactionMeta as bytes.
+func TxMetaEncode(meta *TransactionMeta) []byte {
+	b := new(bytes.Buffer)
+
+	// If the SignatureHashType is not explicitly defined, then it uses EIP155.
+	sighashType := meta.SignatureHashType
+	if sighashType == nil {
+		sighashType = &SighashEIP155
```

Comment:
In this and other cases, it seems dangerous to be using the address of `SighashEIP155`, as accidental modification would affect every reference. The other option is instantiating a `SignatureHashType` pointer and setting its value to `SighashEIP155` and/or creating a factory function to do that.

Reply:
Now that the `SignatureHashType` is no longer part `Transaction.data`, it could also be made into a `const`. I originally went down that path but I couldn't think of a backwards compatible way with the RLP serialization at the time. This would mean that it would no longer be a pointer throughout the codebase so we wouldn't need to worry about cases when it is `nil`. All `NewTransaction` calls would have to be updated. 

I agree with your concern, the factory function idea is good. Something like this?

```golang
func getSighashEIP155() *SignatureHashType {
    return &SighashEIP155
}
```

Reply:
So I pushed up a fix that uses a factory function like above. I'm back and forth on turning it into a `const`

Reply:
Pushed up a commit that uses `const`s for the sighash types